### ISSUE TITLE
您好，我对您的研究成果非常感兴趣，请问您可以提供一下可视化代码吗

### ISSUE THREAD

**Issue by Harry-Bai** [xcyao00/ResAD#2]

**Comment by xcyao00** [comment]
For visualizing anomaly detection score maps, you can use the code in **visualizer.py**.

For visualizing initial and residual feature tsne distribution, you can try the following code:

```python
def plot_initail_and_res_features():
    data = np.load('features.npz')
    ini_feats = data['initial_feats']
    res_feats = data['residual_feats']
    mask = data['mask']
    tsne = manifold.TSNE(n_components=2, init='pca', random_state=0)
    ini_feats = tsne.fit_transform(ini_feats).T
    tsne = manifold.TSNE(n_components=2, init='pca', random_state=0)
    res_feats = tsne.fit_transform(res_feats).T


    fig, axs = plt.subplots(1, 2, figsize=(6, 3))
    axs[0].scatter(ini_feats[0, mask == 0], ini_feats[1, mask == 0], s=10, c='mediumseagreen', cmap=plt.cm.rainbow)
    axs[0].scatter(ini_feats[0, mask == 1], ini_feats[1, mask == 1], s=10, marker='^', c='orange', cmap=plt.cm.rainbow)
    axs[1].scatter(res_feats[0, mask == 0], res_feats[1, mask == 0], s=10, c='mediumseagreen', cmap=plt.cm.rainbow)
    axs[1].scatter(res_feats[0, mask == 1], res_feats[1, mask == 1], s=10, marker='^', c='orange', cmap=plt.cm.rainbow)

    axs[0].set_xticks([])
    axs[0].set_yticks([])
    # axs[1].set_xticks([])
    # axs[1].set_yticks([])

    plt.savefig('tsne.jpg', dpi=1000)
    plt.show()
```

Note that you need first plug some code snippet in the train script to save the initial and residual features, and the corresponding masks (in the above code, the features are saved in the features.npz). Other feature visualizations are also similar to the above.

**Comment by Harry-Bai** [comment]
Thank you very much. Wishing you a smooth scientific research

**Comment by Harry-Bai** [comment]
My coding ability is a bit weak, can you teach me how to call visualizer.py？
